rfcs/006-package-management.md

```diff
@@ -0,0 +1,243 @@
+be named `package.ncl`, and they are found by searching up from the file being
+evaluated. That is, when the user invokes `nickel export path/to/foo.ncl`, we
+look for a manifest at `path/to/package.ncl` and then at `path/package.ncl`, and
+so on.
+
+The manifest file format is defined by the contract `std.package.Manifest`,
+while is defined as
+
+```nickel
+{
+  name | String,
+  version | Semver,
+  nickel-version | Semver,
+
+  dependencies
+    | { _:
+      [|
+        'Path String,
+        'Git { url | String, branch | optional | String, rev | optional | String },
+        'Index { name | String, version | Semver },
+      |]
+      }
+    | default
+    = {},
+}
+```
+
+So an example manifest might look like
+
+```nickel
+{
+  # ...
+  dependencies = {
+    foo = 'Index { package = "github/tweag/foo", version = "1.2.0" }
+    bar = 'Path "../my-bar",
+  }
+} | std.package.Manifest
+```
+
+### Alternative: inline dependencies
+
+Nix and Dhall allow for importing dependencies dynamically, using things like
+`fetchGit`. There was some discussion
+[here](https://github.com/tweag/nickel/issues/329#issuecomment-967372858)
+on the advantages and disadvantages of inline imports.
+
+### Alternative: toml manifest
+
+Maybe the manifest should be in some plain-data format like toml. This would
+be easier to modify programmatically.
+
+### Alternative: shorthand for registry imports
+
+Cargo allows a shorthand like
+
+```toml
+    "github/tweag/foo" = "1.2.0"
+```
+
+instead of
+
+```nickel
+    foo = 'Index { package = "github/tweag/foo", version = "1.2.0" }
+```
+
+Since we expect registry imports to be the common case, maybe it's worth having
+a shorthand?
+
+### Question: manifest file name
+
+The name `package.ncl` was chosen to be similar to npm's `package.json` or
+stack's `package.yaml`. One problem with this is that it could be confused for
+a nickel source file. Another possibility would be to use an extension-based
+name like `<package-name>.cabal`. Or just a stranger name that's less likely to
+conflict with something real.
+
+## Import statements
+
+The manifest file assigns a name to each dependency; to import the dependency
+named `foo` you simply write `import foo`. That is, an `import` statement either
+takes a string in quotes -- in which case it imports a path -- or an identifier
+without quote -- in which case it imports a package.
+
+The `import foo` expression evaluates to the contents of `main.ncl` in `foo`'s
+root directory.
+
+### Question: other entry points?
+
+We've hardcoded `main.ncl` as the entry point of every package, but what if
+they want to expose multiple entry points? For example, node allows a package's
+manifest to specify the entry point(s). This is probably not very important to
+support, as you can just put
+
+```nickel
+{
+  other = import "./other.ncl",
+  blah = import "./blah.ncl",
+}
+```
+
+in your package's `main.ncl`, to provide "other" and "blah" as other entry points.
```

Comment:
Another possibility would be to make all the top-level files separate entry points. But since we don't have public/private modifiers, it might prove annoying, as it makes it harder to hide internal modules, so to speak.

Reply:
Sounds reasonable, I'll add it as an alternative. I think it isn't too hard to hide private modules -- just put them in a subdirectory.